internal/services/stock_exchange/algorithms.go

```diff
@@ -0,0 +1,340 @@
+	config.DB.Raw("SELECT * FROM stocks WHERE stock_name = ?", stock).Scan(&currentStock)
+	currentStock.HighPrice = int(math.Max(float64(ltp), float64(currentStock.HighPrice)))
+	currentStock.LowPrice = int(math.Min(float64(ltp), float64(currentStock.LowPrice)))
+	currentStock.LTP = ltp
+	currentStock.UpdatedAt = time.Now()
+	//fmt.Println(currentStock)
+	config.DB.Exec("UPDATE stocks SET ltp = ? ,high_price = ? ,low_price = ? , updated_at = ?  WHERE stock_ticker_symbol = ? ", ltp, currentStock.HighPrice, currentStock.LowPrice, currentStock.UpdatedAt, stock)
+	//config.DB.Table("stocks").Save(&currentStock)
+}
+
+func BuyOrderMatchingAlgo(buyOrderBody OrderRequest, resp OrderResponse, id int) {
+
+	var sellBook []model.SellOrderBook
+	// db lock
+	err := config.DB.Raw("SELECT * FROM sell_order_book WHERE stock_ticker_symbol = ?  ORDER BY order_price ASC,created_at ASC ", buyOrderBody.StockName).Scan(&sellBook).Error
+	if err != nil {
+		resp.Status = "Db Fetch failed, order cancelled"
+		config.DB.Exec("DELETE FROM buy_order_book WHERE id = ?", id)
+		orderResponse <- resp
+		return
+	}
+	if len(sellBook) == 0 {
+		// abort with message not enough shares
+		resp.Status = "Db Fetch failed, order cancelled"
+		orderResponse <- resp
+		config.DB.Exec("DELETE FROM buy_order_book WHERE id = ?", id)
+		return
+	}
+	SharesAvailable := 0
+	for _, elem := range sellBook {
+		SharesAvailable += elem.OrderQuantity
+	}
+
+	if SharesAvailable < buyOrderBody.Quantity {
+		resp.Status = "Shares not available, order cancelled"
+		resp.Message = "Shares not available"
+		config.DB.Exec("DELETE FROM buy_order_book WHERE id = ?", id)
+		orderResponse <- resp
+		return
+	}
+	totalCost := 0
+	ltp := 0
+	orderedQuantity := buyOrderBody.Quantity
+	if buyOrderBody.OrderType == "Limit" {
+		//if int(sellBook[0].OrderPrice) <= price {
+		for _, elem := range sellBook {
+
+			if elem.OrderPrice <= buyOrderBody.LimitPrice {
+				if elem.OrderQuantity > buyOrderBody.Quantity {
+					ltp = elem.OrderPrice
+					config.DB.Exec("UPDATE sell_order_book SET order_quantity = ? , updated_at = ? WHERE id = ? ", elem.OrderQuantity-buyOrderBody.Quantity, time.Now(), elem.ID)
+					totalCost += elem.OrderPrice * buyOrderBody.Quantity
+					buyOrderBody.Quantity -= buyOrderBody.Quantity
+					UpdateLTP(ltp, buyOrderBody.StockName)
+					break
+				} else {
+					ltp = elem.OrderPrice
+					UpdateLTP(ltp, buyOrderBody.StockName)
+					totalCost += elem.OrderPrice * elem.OrderQuantity
+					buyOrderBody.Quantity -= elem.OrderQuantity
+					config.DB.Exec("DELETE FROM sell_order_book WHERE id = ?", elem.ID)
+					//config.DB.Delete(&model.SellOrderBook{}, elem.ID)
+				}
+			} else {
+				// add to pending order list in order book
+				config.DB.Exec("UPDATE buy_order_book SET order_quantity = ? , updated_at = ? WHERE id = ?", buyOrderBody.Quantity, time.Now(), id)
+				time.Sleep(30 * time.Second)
+				//Recursive call :
+				BuyOrderMatchingAlgo(buyOrderBody, resp, id)
```

Comment:
Can we run a simple infinite for loop with some sleep? instead of recursion.

Reply:
Don't think that is possible here. Might have to change entire code to do that.